packages/myst-cli/src/project/fromTOC.ts

```diff
@@ -31,156 +32,44 @@ import { globSync } from 'glob';
- * }
- */
-export function objFromPathParts(parts: string[], obj?: Record<string, any>) {
-  if (!obj) obj = {};
-  const [part, ...remainingParts] = parts;
-  if (!part) return obj;
-  if (!obj[part]) obj[part] = {} as Record<string, any>;
-  obj[part] = objFromPathParts(remainingParts, obj[part]);
-  return obj;
-}
+export function comparePaths(a: string, b: string): number {
+  const aDirName = dirname(a);
+  const bDirName = dirname(b);
+  // Same directory?
+  if (aDirName === bDirName) {
+    const aBaseName = basename(a);
+    const bBaseName = basename(b);
 
-/**
- * Take glob pattern and create nested pseudo-directory structure
- *
- * For example, given a pattern that resolves to:
- *
- * matches = [
- *   'one/two/three',
- *   'one/two/four',
- *   'five',
- * ]
- *
- * This builds:
- *
- * {
- *   one: {
- *     two: {
- *       three: {},
- *       four: {},
- *     },
- *   },
- *   five: {},
- * }
- */
-export function patternToDirectoryStructure(
-  pattern: string,
-  path: string,
-  opts?: Omit<Parameters<typeof globSync>[1], 'withFileTypes'>,
-) {
-  const matches = globSync(pattern, { cwd: path, ...opts });
-  const dirStructure: Record<string, any> = {};
-  matches.forEach((match) => {
-    // Glob always uses '/' separator, not filesystem separator
-    const fileParts = match.split('/');
-    objFromPathParts(fileParts, dirStructure);
-  });
-  return dirStructure;
-}
+    const aIsIndex = DEFAULT_INDEX_WITH_EXT.includes(aBaseName);
+    const bIsIndex = DEFAULT_INDEX_WITH_EXT.includes(bBaseName);
 
-/**
- * Take "directory structure" object and return corresponding toc entries
- *
- * The resulting entries are only FileEntries and ParentEntries. Similar
- * to the implicit project ordering, files come before parents.
- *
- * For example, given:
- *
- * dirStructure = {
- *   one: {
- *     two: {
- *       three: {},
- *       four: {},
- *     },
- *   },
- *   five: {},
- * }
- *
- * This builds:
- *
- * [
- *   { file: 'five' },
- *   {
- *     title: one,
- *     children: [
- *       {
- *          title: two,
- *          children: [
- *            { file: 'four' },
- *            { file: 'three' },
- *          ],
- *       },
- *     ],
- *   },
- * ]
- */
-export function directoryStructureToFileEntries(
-  dirStructure: Record<string, any>,
-  path: string,
-  ignore: string[],
-): (FileEntry | ParentEntry)[] {
-  let files = Object.keys(dirStructure)
-    // .map((file) => join(path, file))
-    .filter((file) => !ignore || !ignore.includes(join(path, file)))
-    .sort(sortByNumber);
-  let indexFile: string | undefined;
-  DEFAULT_INDEX_WITH_EXT.forEach((index) => {
-    if (indexFile) return;
-    indexFile = files.find((file) => file.toLowerCase() === index);
-  });
-  if (indexFile) {
-    files = [indexFile, ...files.filter((file) => file !== indexFile)];
+    // Are both a and b index files, or neither?
+    if (aIsIndex === bIsIndex) {
+      return aBaseName.localeCompare(bBaseName, undefined, {
```

Comment:
> This is an extreme edge case, but if there are two index files, index.md and index.ipynb, the .ipynb file will come first. It also means all INDEX files will be first. This is different from the implicit toc behavior where (1) index preference prioritizes .md over other formats and (2) once an index file is found, the other index files are just treated as normal files...

Yes -- although I hadn't thought about this regression, I think it's worth keeping for simplicity; then we can keep our sorting logic to be local rather than stateful?